be/src/exec/pipeline/sink/connector_sink_operator.cpp

```diff
@@ -0,0 +1,154 @@
+        }
+    }
+#ifndef BE_TEST
+    Operator::close(state);
+#endif
+}
+
+bool ConnectorSinkOperator::need_input() const {
+    if (_no_more_input) {
+        return false;
+    }
+
+    while (!_add_chunk_future_queue.empty()) {
+        // cannot accept chunk if any add_chunk_future is not ready
+        if (!is_ready(_add_chunk_future_queue.front())) {
+            return false;
+        }
+        if (auto st = _add_chunk_future_queue.front().get(); !st.ok()) {
+            LOG(WARNING) << "cancel fragment: " << st;
+            _fragment_context->cancel(st);
+        }
+        _add_chunk_future_queue.pop();
+    }
+
+    return true;
+}
+
+bool ConnectorSinkOperator::is_finished() const {
+    if (!_no_more_input) {
+        return false;
+    }
+
+    while (!_add_chunk_future_queue.empty()) {
+        // unfinished if any add_chunk_future future is not ready
+        if (!is_ready(_add_chunk_future_queue.front())) {
+            return false;
+        }
+
+        if (auto st = _add_chunk_future_queue.front().get(); !st.ok()) {
+            LOG(WARNING) << "cancel fragment: " << st;
+            _fragment_context->cancel(st);
+        }
+        _add_chunk_future_queue.pop();
+    }
+
+    while (!_commit_file_future_queue.empty()) {
+        // unfinished if any commit_file_future future is not ready
+        if (!is_ready(_commit_file_future_queue.front())) {
+            return false;
+        }
+
+        auto result = _commit_file_future_queue.front().get();
+        _commit_file_future_queue.pop();
+
+        if (auto st = result.io_status; st.ok()) {
+            // invoke callback if file commit succeed
+            _connector_chunk_sink->callback_on_success()(result);
+        } else {
+            LOG(WARNING) << "cancel fragment: " << st;
+            _fragment_context->cancel(st);
+        }
+        _rollback_actions.push(std::move(result.rollback_action));
+    }
+
+    DCHECK(_add_chunk_future_queue.empty());
+    DCHECK(_commit_file_future_queue.empty());
+    return true;
+}
+
+Status ConnectorSinkOperator::set_finishing(RuntimeState* state) {
+    _no_more_input = true;
+    auto future = _connector_chunk_sink->finish();
+    _enqueue_futures(std::move(future));
+    return Status::OK();
+}
+
+bool ConnectorSinkOperator::pending_finish() const {
+    return !is_finished();
+}
+
+Status ConnectorSinkOperator::set_cancelled(RuntimeState* state) {
+    _is_cancelled = true;
+    return Status::OK();
+}
+
+StatusOr<ChunkPtr> ConnectorSinkOperator::pull_chunk(RuntimeState* state) {
+    CHECK(false) << "ConnectorSinkOperator::pull_chunk";
+    __builtin_unreachable();
+}
+
+Status ConnectorSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
+    ASSIGN_OR_RETURN(auto future, _connector_chunk_sink->add(chunk));
+    _enqueue_futures(std::move(future));
+    return Status::OK();
+}
+
+void ConnectorSinkOperator::_enqueue_futures(connector::ConnectorChunkSink::Futures future) {
```

Comment:
param name: connector::ConnectorChunkSink::Futures future, suggest to use futures.
Futures' fields add_chuck_future and commit_file_future, also suggest to use plurals.

Reply:
renamed.